### ISSUE TITLE
Replace direct require() calls with an indirection function

### ISSUE THREAD

**Issue by parisholley** [mikro-orm/mikro-orm#3743]
**Is your feature request related to a problem? Please describe.**
Many new JS stacks use some type compiler (swc, esbuild, babel) and will often have some setting enabled that will attempt to in-line all code that is included via require(). That is problematic for this project since things like database drivers are only included if needed.

**Describe alternatives you've considered**
You can try to figure out how to exclude or mock the module name but for some projects (such as remix), the compiler stack isn't exposed to you unless you patch the project.

**Comment by B4nan** [comment]
I don't think we have require calls that cannot be already gotten around (e.g. for drivers you can pass the driver class via `driver` option instead of the `type` option). You can already use MikroORM with esbuild/swc, as long as you are fine with not using decorators for metadata.

Or is this about removing the require calls, so you don't need special handling for them (e.g. as in webpack)? We need defaults that work universally, so I don't think that's doable in near future.

**Comment by B4nan** [comment]
Looking at the PR, I believe that is the exact way to break things. This won't work in ESM project at all, if I'm not mistaken - will try that tomorrow, I have one ESM example project.

FYI we also have `Utils.requireFrom` and `Utils.tryRequire`, those are tested in various types of project.

**Comment by parisholley** [comment]
> as long as you are fine with not using decorators for metadata

decorators work with swc, you just have to define `type` for every property and i have a `registerMikro` function for gathering classes in an array and passing to my mikro init logic

> that is the exact way to break things

i'm not familiar enough with ESM to say

> we also have `Utils.requireFrom` and `Utils.tryRequire`

alright, i can change if you think this isn't a breaking change

**Comment by parisholley** [comment]
take this comment with a grain of salt as I am biased :) but using webpack to bundle server-side node_modules seems like a smell? i could get trying tree-shake dependencies or maybe having a single file to say, deploy to a serverless function provider, but even in SSR situations node_modules i find to be separate most of the time and a lot of projects meant for node aren't tree-shakable to begin with. 

@B4nan on flip side, obviously a breaking change, could we not just have the user pass an instance/class/etc to config instead of driver "names" (same thing with migrations), and this problem goes away?

**Comment by B4nan** [comment]
Well, I don't have any opinion about this, never used webpack for it, so not interested in discussing it in any way (which is why I was tagging Kevin, he introduced the change, I am sure he will have some opinions). But your change is reverting a fix, that's what I care about - you might fix something for yourself, and break something that worked for others. If we are to do it, we need a working guide for webpack. I dont care if it will suggest doing X or Y, but it needs to be correct and working.

Note that you can already pass a driver class instead of using the `type` option. The option was added because that way you can set your driver via env vars, so we need some kind of dynamic require in place.

**Comment by thekevinbrown** [comment]
Hey @B4nan and @parisholley! I'm still around and we do still use Mikro.

We deploy our server side code to Lambda. It's really nice to have a single file which is the entire application and all of its dependencies at a particular version which we then just throw at a stock standard Lambda image with node. This makes it easy to deploy / roll back / etc. We use webpack currently (but are using ESBuild in dev already, so will likely be doing that in prod soon) to achieve this.

While there are other methods (like layers) which we could use to manage the application in more of an un-bundled way, this would increase complexity, and we've been in production like this for a couple of years at this point.

Most compilers / bundlers that can inline `require()` contents also expose options so you can control this behaviour, either by stubbing out modules that aren't found, or by simply rewriting the statement to `null`. Absent changes in Mikro that'd be my recommended approach so as not to break people who use webpack / ESBuild.

In general my preference would be to get Mikro out of the `require()` game here altogether by removing the ability to specify a name and just having the consumer of the library pass in an instance of whatever platform they want to the configuration. That way they can get it from wherever they want.

**Comment by parisholley** [comment]
@thekevinbrown yeah, figured lambda bundling was backstory here, i always use layers to avoid fighting bundle-unfriendly projects but it does add another layer to DevOps.

> Most compilers / bundlers that can inline require() contents also expose options 

suppose it depends on what world of JS you live in, SSR heavy frameworks don't expose very much and you'd have to dig into their code to patch their vite/esbuild/swc/etc pipelines. at the moment, i have to choose between hacking their framework or hacking mikro. 

I agree that the only real solution here is to remove the requires out all together (in any capacity) and require the user to include them and that would work with any bundler system. as it stands, only webpack users in your situation are benefiting from the "fix", but this would be breaking change for v5.

options I see with v5:
- add every require'd module to my package.json, probably not of much consequence
- continue with patching those lines out (PITA with upgrades)
- introduce a new main/index (eg: `@mikro/core/v6`) that uses an alternate version without requires and potentially make it default in next major

**Comment by B4nan** [comment]
We can surely get rid of the `type` option, at least to a certain degree - v5.5 added the driver-specific `MikroORM` exports that specify the driver type by class. As mentioned, the `type` was mainly there for those who want to use env vars only, but again, we have driver-specific `defineConfig` helpers that do the same, so we don't really need the `type` option if those new constructs are used. 

There are more places that require things dynamically, but I guess that is mostly for the external stuff, that can be always used in other ways - we could keep something like `Utils.tryRequire` or `Utils.requireFrom` in there to preserve the API for people who use things "normally", if not, just use the package directly (e.g. import the entity generator from its package). 

Totally agree that it would be best to get rid of it all. We will still need some dynamic require for entities, migrations, etc, but that's again all opt-in.

I don't think we should care much about this in v5, it does not seem to be possible to fix it in a non-breaking way, and I'd rather not come up with some alternative exports here (it would be really just a partial solution). I'd like to start the development rather soon, planning maybe one or two more feature releases, then we can branch 5.x and keep master for v6, backporting fixes to 5.x when appropriate, keeping all features to v6.

**Comment by B4nan** [comment]
On the other hand, I must say I really like the new shortcuts (e.g. `orm.schema`)... Maybe optional dependencies would be more appropriate here? I used peer deps because as those will be installed by user, and I believe NPM installs optional deps by default, so it wouldn't be really a solution.

Or make it work the other way, users would "install extensions" in the ORM config, which would basically provide (safe) implementation for those methods. Every extension like that could export a `register` method:

```ts
import { defineConfig } from '@mikro-orm/postgres';
import { register as migrator } from '@mikro-orm/migrations';
import { register as entityGenerator } from '@mikro-orm/entity-generator';
import { register as seeder } from '@mikro-orm/seeder';

export default defineConfig({
  dbName: 'test',
  extensions: [migrator, entityGenerator, seeder],
});
```

Also tagging @merceyz for some insights around yarn PNP, if we are to redesign this in v6, I'd want to be compatible with that too. But I believe we should be fine if we get rid of all the static `require` calls, right?

**Comment by B4nan** [comment]
I think #3773 should do it. Once we start with v6 development, we can just remove all of the static `require`s as well as the `type` option and we should be good to go. I will also probably do the reexport of all core/knex classes in the driver package so users shouldn't need to use the `core` package in their imports at all, to reduce confusion.

**Comment by merceyz** [comment]
> Also tagging @merceyz for some insights around yarn PNP, if we are to redesign this in v6, I'd want to be compatible with that too. But I believe we should be fine if we get rid of all the static require calls (and optional peer dependencies too I guess), right?

Yeah, that seems fine 👍

**Comment by B4nan** [comment]
I hope #3814 should cover this. v6 branch is set up to publish v6 canary builds, so once it's merged, we can start testing how it actually behaves across the different frameworks and environments. I removed most of the peer dependencies that caused cycles, the `type` option (kept the `MIKRO_ORM_TYPE` env var for now, with a `createRequire` used under the hood).

**Comment by B4nan** [comment]
#3814 is merged to https://github.com/mikro-orm/mikro-orm/tree/v6 and published as `6.0.0-dev.6`, so closing. Let's test it and open new issues if there are more changes needed.
